components/brave_wallet/browser/internal/hd_key.cc

```diff
@@ -430,6 +431,35 @@ std::vector<uint8_t> HDKey::GetPublicKeyFromX25519_XSalsa20_Poly1305() const {
   return public_key;
 }
 
+std::vector<uint8_t> HDKey::DecryptCipherFromX25519_XSalsa20_Poly1305(
+    const std::string& version,
+    const std::vector<uint8_t>& nonce,
+    const std::vector<uint8_t>& ephemeral_public_key,
+    const std::vector<uint8_t>& ciphertext) const {
+  // Only x25519-xsalsa20-poly1305 is supported by MM at the time of writing
+  if (version != "x25519-xsalsa20-poly1305")
+    return std::vector<uint8_t>();
+  if (nonce.size() != crypto_box_curve25519xsalsa20poly1305_tweet_NONCEBYTES)
+    return std::vector<uint8_t>();
+  if (ephemeral_public_key.size() !=
+      crypto_box_curve25519xsalsa20poly1305_tweet_PUBLICKEYBYTES)
+    return std::vector<uint8_t>();
+  if (private_key_.size() !=
+      crypto_box_curve25519xsalsa20poly1305_tweet_SECRETKEYBYTES)
+    return std::vector<uint8_t>();
+
+  std::vector<uint8_t> padded_ciphertext = ciphertext;
+  padded_ciphertext.insert(padded_ciphertext.begin(), 16, 0);
+  std::vector<uint8_t> padded_plaintext(ciphertext.size() + 32);
+  if (crypto_box_open(padded_plaintext.data(), padded_ciphertext.data(),
```

Comment:
> Nonce checking to see if it's a super low number (like less than 10) and fire a debug warning off. The "X" in this crypto construct implies that the nonce is meant to be a random 192 bit nonce so a super lower nonce implies the sender is picking bad nonces (usually picking a counter construct). Since there's realistically nothing we can do by this point where the message has already been sent over the wire the warning would just be about alerting the devs sending the message they're doing something wrong. If this isn't immediately easy we can just ignore it since it's an extremely defensive security posture.

I could see that for an encrypt API but as you say the decrypt API it's already too late. So it could just lead to something that needs to be decrypted not being able to decrypt. I'd prefer not to implement this one unless there's a strong preference.

> zeroing out the private keys once we're done with using it. I'm not sure this is the best spot (we could also be already doing it since I didn't dig much further into this yet) to do this though since it looks like the private_key memory isn't being allocated here and the best technique I've found for this practice is to zero out in the same function as where you allocate for readability of the code and ease of checking.

I believe this is covered by the auto lock timer that we have. At least it's the same as all other keyring usage that we have. cc @darkdh for an opinion though.

Reply:
It's already zero out when wallet get locked or browser exit
https://github.com/brave/brave-core/blob/fbe8d66420ca557cda922ca7c97a65e0d1b6a893/components/brave_wallet/browser/internal/hd_key.cc#L113

Reply:
Thanks for link to how we zero keys and explaining our approach as well. As for the nonce yeah no strong preference here so definitely good with just leaving it as is.

I'll mark this one as resolved.